### PR TITLE
Rename "towers" to "inverse sequential diagrams"

### LINKED ISSUE #965: Sync up nomenclature between towers and cotowers
See https://github.com/UniMath/agda-unimath/pull/919#issuecomment-1826941011

### PR DESCRIPTION
**Pull request by fredrik-bakke**
This is to make the terminology consistent with its dual, sequential diagrams(/cotowers). Resolves #965.

### CODE REVIEW

File: src/foundation/sequential-limits.lagda.md [hunk 1]
```diff
@@ -210,45 +221,55 @@ module _
...
-### A cone over a tower is equal to the value of `cone-map-tower` at its own gap map
+### A cone over an inverse sequential diagram is equal to the value of `cone-map-inverse-sequential-diagram` at its own gap map
 
 ```agda
 module _
-  {l1 l2 : Level} (A : tower l1) {X : UU l2}
+  {l1 l2 : Level} (A : inverse-sequential-diagram l1) {X : UU l2}
   where
 
   htpy-cone-up-pullback-standard-sequential-limit :
```
Comment (VojtechStep) [inline src/foundation/sequential-limits.lagda.md:267]: What does this have to do with pullbacks?

File: src/foundation/transfinite-cocomposition-of-maps.lagda.md [hunk 2]
```diff
@@ -7,17 +7,18 @@ module foundation.transfinite-cocomposition-of-maps where
...
 ```
 
 </details>
 
 ## Idea
 
-Given a [tower](foundation.towers.md) of maps, i.e. a certain infinite
-[sequence](foundation.dependent-sequences.md) of maps `fₙ`:
+Given an [inverse sequential diagram](foundation.inverse-sequential-diagrams.md)
+of maps, i.e. a certain infinite [sequence](foundation.dependent-sequences.md)
```
Comment (VojtechStep) [inline src/foundation/transfinite-cocomposition-of-maps.lagda.md:20]: "inverse sequential diagram of maps" isn't defined anywhere, only diagrams of types; is that intentional?

**Review by VojtechStep — COMMENTED** [review]
Apart from mostly cosmetic requests, this LGTM 👌

File: src/foundation/inverse-sequential-diagrams.lagda.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+```agda
+sequence-map-inverse-sequential-diagram : {l : Level} → (ℕ → UU l) → UU l
+sequence-map-inverse-sequential-diagram A = (n : ℕ) → A (succ-ℕ n) → A n
+
+inverse-sequential-diagram : (l : Level) → UU (lsuc l)
+inverse-sequential-diagram l =
+  Σ (ℕ → UU l) (sequence-map-inverse-sequential-diagram)
+
+type-inverse-sequential-diagram :
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/inverse-sequential-diagrams.lagda.md:55]: Oh, I had forgotten about that issue.

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/inverse-sequential-diagrams.lagda.md:55]: I agree that this PR should resolve that issue. Do you know if there is nothing else to change? Otherwise, I'll go through the files systematically.

File: src/foundation/equivalences-inverse-sequential-diagrams.lagda.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+open import foundation-core.homotopies
+open import foundation-core.identity-types
+open import foundation-core.torsorial-type-families
+```
+
+</details>
+
+## Idea
+
+An **equivalence of inverse sequential diagrams** `A ≃ B` is a commuting diagram
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/equivalences-inverse-sequential-diagrams.lagda.md:33]: Egbert recently commented that he does not like the practice of inserting links in terms that are being defined, so I'm not sure that a link should be inserted here.

File: src/foundation/sequential-limits.lagda.md (hunk 1 above)
Comment (fredrik-bakke) [reply src/foundation/sequential-limits.lagda.md:267]: That's an old mistake. Thanks for spotting it!

File: src/foundation/transfinite-cocomposition-of-maps.lagda.md (hunk 2 above)
Comment (fredrik-bakke) [reply src/foundation/transfinite-cocomposition-of-maps.lagda.md:20]: That's also a mistake. Thanks!

File: src/foundation/inverse-sequential-diagrams.lagda.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (VojtechStep) [reply src/foundation/inverse-sequential-diagrams.lagda.md:55]: We also discussed whether we should have the shapes of coherences in the names of accessors on (co)cones, like if there should be `coherence-cone-inverse-sequential-diagram` or `coherence-triangle-cone-inverse-sequential-diagram`. I've come around to think that the shapes are redundant, so I didn't suggest changing it here.

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/inverse-sequential-diagrams.lagda.md]: I think I need a little more time to think the issue over with regards to mentioning the shape of the diagram.

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/inverse-sequential-diagrams.lagda.md]: Okay, I ended up just removing `triangle`. There's still the question of whether we should use `up`. Currently, I think it's okay to use `up` (and `dup`), like we do `coh` vs. `coherence`.

**Comment by fredrik-bakke** [comment]
Alright, I think this one is ready for merging.

File: src/foundation/inverse-sequential-diagrams.lagda.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/inverse-sequential-diagrams.lagda.md]: I could note, though, that there generally shouldn't generally be a naming conflict forcing us to use `up` over `universal-property`, since we use `standard-pullback` and so on for the standard limit. I still like `up` because `universal-property` is very verbose.

**Comment by EgbertRijke** [comment]
Excellent, I’ll have a look at this today

File: src/foundation-core/pullbacks.lagda.md [hunk 5]
```diff
@@ -170,15 +170,14 @@ module _
   where
 
   abstract
-    universal-property-pullback-standard-pullback :
+    up-standard-pullback :
```
Comment (EgbertRijke) [inline src/foundation-core/pullbacks.lagda.md:173]: I've been advocating for a long time to write `up` as `universal-property`. It reads a lot better in my opinion, and I'm not fond of this reversal

Comment (fredrik-bakke) [reply src/foundation-core/pullbacks.lagda.md:173]: The purpose of this change is to make the names consistent with each other. If we want to replace `up` with `universal-property`, then that should be part of a different pull request.

Comment (fredrik-bakke) [reply src/foundation-core/pullbacks.lagda.md:173]: To be clear, this change is not a statement of preference.

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+open import foundation-core.function-types
+```
+
+</details>
+
+## Idea
+
+Given a ([dependent](foundation.dependent-inverse-sequential-diagrams.md))
+[inverse sequential diagram](foundation.inverse-sequential-diagrams.md) `A`, we
```
Comment (EgbertRijke) [inline src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: Linking 3/4 of the phrase "dependent inverse sequential diagram" to something else looks misleading to me. I'd prefer if the entire phrase just linked to the file about dependent inverse sequential diagrams.

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+can extract the **composite map from `Aₙ₊ᵣ` to `Aₙ`**.
```
Comment (EgbertRijke) [inline src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:26]: We could use the `{{#concept }}` macro here too

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
```
Comment (EgbertRijke) [inline src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:27]: This idea merely states that we can extract the composite, but it doesn't make clear what that composite is, what type it has, or what diagram it fits in. The reader will not be able to understand the idea from what is written here.

Comment (EgbertRijke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:27]: In particular, I don't really understand what you mean:)

File: src/foundation-core/pullbacks.lagda.md (hunk 5 above)
Comment (fredrik-bakke) [reply src/foundation-core/pullbacks.lagda.md:173]: Although I personally don't mind `up`. It reads similar to our usage of `coh`, and avoids a lot of verbosity. It also makes `up-pullback` the exact same length as `is-pullback`.

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+## Definitions
+
+### Composite maps in inverse sequential diagrams
+
+```agda
+comp-map-inverse-sequential-diagram :
+  {l : Level} (A : inverse-sequential-diagram l) (n r : ℕ) →
+  family-inverse-sequential-diagram A (n +ℕ r) →
```
Comment (EgbertRijke) [inline src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:35]: One could argue that the type of this composite should be `family-inverse-sequential-diagram A m -> family-inverse-sequential-diagram A n` where an identification `m = n + r` is given as a parameter. This would avoid complications down the line when something is not judgmentally equal to the correct sum of natural numbers (eg associativity is going to cause headaches under this setup)

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md (hunk 7 above)
Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:26]: The purpose of this PR is only to make the necessary changes to resolve #965. Therefore, I am a little reluctant to make the changes you ask for here. However, I can make a follow-up PR where I improve some of the phrasing for these files as there has been a recent comment on Discord complaining about being confused about these files as well.

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  family-inverse-sequential-diagram A n
+comp-map-inverse-sequential-diagram A n zero-ℕ = id
+comp-map-inverse-sequential-diagram A n (succ-ℕ r) =
+  comp-map-inverse-sequential-diagram A n r ∘
+  map-inverse-sequential-diagram A (n +ℕ r)
+```
+
+### Composite maps in dependent inverse sequential diagrams
```
Comment (EgbertRijke) [inline src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:43]: This could easily have its own file

File: src/foundation/cones-over-inverse-sequential-diagrams.lagda.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+open import foundation-core.identity-types
+open import foundation-core.torsorial-type-families
+open import foundation-core.whiskering-homotopies
+```
+
+</details>
+
+## Idea
+
+A **cone over an
```
Comment (EgbertRijke) [inline src/foundation/cones-over-inverse-sequential-diagrams.lagda.md:34]: We could use the `{{#concept }}` macro here

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md (hunk 6 above)
Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: And do other refactoring related to inverse sequential diagrams

File: src/foundation-core/pullbacks.lagda.md (hunk 5 above)
Comment (EgbertRijke) [reply src/foundation-core/pullbacks.lagda.md:173]: > The purpose of this change is to make the names consistent with each other. If we want to replace `up` with `universal-property`, then that should be part of a different pull request.

This could perhaps be relevant if I hadn't already advocated for `universal-property` for a long time and if we didn't have `universal-property` in full in lots of places already. I oppose this reversal back to `up`.

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md (hunk 6 above)
Comment (EgbertRijke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: Ah ok, but this PR goes beyond #965 in that it creates many new files. Why not take care of the prose of those new files as well?

I don't think there's a rush with getting this PR merged, and I'd like to give it a proper review. In particular on all the lines of code that show as green here on github there are lots of things that I think can be improved and I would still like to comment on them.

File: src/foundation-core/pullbacks.lagda.md (hunk 5 above)
Comment (fredrik-bakke) [reply src/foundation-core/pullbacks.lagda.md:173]: That is going to make the standard pullback the only standard (co)limit in the library that uses `universal-property` in full for this entry.

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md (hunk 6 above)
Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: No, it doesn't create any new files. It just renames old ones.

File: src/foundation-core/pullbacks.lagda.md (hunk 5 above)
Comment (EgbertRijke) [reply src/foundation-core/pullbacks.lagda.md:173]: In spans I'm renaming all the `up`-s in synthetic homotopy theory. It won't be like that for long. I really don't like this change Fredrik.

Comment (fredrik-bakke) [reply src/foundation-core/pullbacks.lagda.md:173]: I just don't see why we can't try to maintain consistency and then make all the changes in one fell swoop instead.

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md (hunk 6 above)
Comment (EgbertRijke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: Ah, that was impossible to see on here. Now I am not sure what to do, because I still think I would like to review all the code that is "changed" according to GitHub.

File: src/foundation-core/pullbacks.lagda.md (hunk 5 above)
Comment (fredrik-bakke) [reply src/foundation-core/pullbacks.lagda.md:173]: But it's okay, I will rename `universal-property-pullback-standard-pullback` to `universal-property-standard-pullback` instead.

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md (hunk 6 above)
Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: You're more than welcome to! I can make use of them in the follow-up

File: src/foundation-core/pullbacks.lagda.md (hunk 5 above)
Comment (EgbertRijke) [reply src/foundation-core/pullbacks.lagda.md]: Thanks so much! Sorry to be so complicated and opinionated:)

Comment (EgbertRijke) [reply src/foundation-core/pullbacks.lagda.md]: BTW over in spans we are exactly trying to be maintain consistency by renaming the rogue `up`'s to `universal-property`. When someone is already doing that work then it would be strange that someone else would make another PR making the opposite change in the name of consistency in a quicker PR.

Comment (fredrik-bakke) [reply src/foundation-core/pullbacks.lagda.md]: I see. I didn't know you were renaming it across the board.

Comment (EgbertRijke) [reply src/foundation-core/pullbacks.lagda.md]: Perhaps not across the board, but you weren't doing that here either and we do have a lot of `universal-property` in the library.

Comment (fredrik-bakke) [reply src/foundation-core/pullbacks.lagda.md]: Which is why I want to defer the resolution to a new PR

File: src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md (hunk 6 above)
Comment (EgbertRijke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: Since you indicate that you don't want to make changes to review comments and I do want to review your code, would you mind if I make a commit into this branch before instead of making a review?

Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: I mean, is it to fix a mistake or to achieve the purpose of the pR? Otherwise, don't you think it is better to open a new PR for the purpose of improving the files' documentation? I am trying to exercise constraining what this PR contains much as possible to avoid going into the same trap I did early in the autumn.

Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: I am more than happy to collaborate on such a PR btw

Comment (EgbertRijke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: It feels so strange to me though, because essentially the suggestion is that we should just merge this without my review. When I've received comments on code that I merely moved around I've always addressed them too. It seems sketchy to me that we should be able have PRs where the submitter says they don't want to resolve review comments because of what they declare the purpose of their PR to be. That should certainly not happen very often.

Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: Well, if the reviewer objects to the purpose of the PR, then I would support blocking such a PR. I still think the contributor deserves some level of protection, otherwise we can exercise power over them to make them implement any additional changes we ask for by keeping their contribution hostage.

Comment (EgbertRijke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: Do you feel that I'm taking your contribution hostage right now?

Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: I have resolved every conversation that pertains to what I essentially see as the purpose of this PR though. See for instance Vojta's previous review.

Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: It seems possible to me that this PR is blocked for reasons outside of what I've intended its purpose to be, because of the additional changes you want to implement, essentially holding me up from getting on with other work. I absolutely do not think there is any malevolence in your motivations, but this is a very annoying situation to be in, which I've found to be quite demotivating in the past. So to avoid experiencing this again, I would very much prefer if the extra changes are implemented in a separate PR.

Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: Sure! You never call. Give me a couple of minutes :)

Comment (fredrik-bakke) [reply src/foundation/composite-maps-in-inverse-sequential-diagrams.lagda.md:25]: I'm available now.

**Comment by EgbertRijke** [comment]
All comments are resolved now
